File: src/main/java/frc/robot/commands/drivebase/AimAtLimelightCommand.java
package frc.robot.commands.drivebase;

import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.subsystems.SwerveSubsystem;
import frc.robot.subsystems.VisionSubsystem;
import frc.robot.subsystems.intake.Intake;
import frc.robot.util.LimelightHelpers;


public class AimAtLimelightCommand extends Command {
  private final SwerveSubsystem swerveSubsystem;
  private final VisionSubsystem visionSubsystem;

  private boolean isRedAlliance;
  double rotateCW, speed;

  public AimAtLimelightCommand(SwerveSubsystem swerveSubsystem, VisionSubsystem visionSubsystem) {
    this.swerveSubsystem = swerveSubsystem;
    this.visionSubsystem = visionSubsystem;

    // each subsystem used by the command must be passed into the
    // addRequirements() method (which takes a vararg of Subsystem)
    addRequirements(this.swerveSubsystem, this.visionSubsystem);
  }


  /**
   * The initial subroutine of a command.  Called once when the command is initially scheduled.
   */
  @Override
  public void initialize() {
    LimelightHelpers.setPipelineIndex("limelight-back", 1);
    isRedAlliance = (DriverStation.getAlliance().get() == DriverStation.Alliance.Red);

    // Take the current angle and see if it's faster to rotate clockwise or counterclockwise to get to 180 degrees
    double currentAngle = swerveSubsystem.getHeading().getDegrees(); // (-180, 180]
    speed = 2.5;
    rotateCW = (currentAngle > 0) ? -speed : speed;

  }

  @Override
  public void execute() {

    if (LimelightHelpers.getTV("limelight-back")) {
      swerveSubsystem.getSwerveDrive().drive(new Translation2d(0, 0), visionSubsystem.limelight_aim_proportional_back() / 5, false, false);
    } else {
      if (swerveSubsystem.getHeading().getDegrees() < 140 && swerveSubsystem.getHeading().getDegrees() > -140)
        swerveSubsystem.getSwerveDrive().drive(new Translation2d(0, 0), rotateCW, false, false);
      else swerveSubsystem.getSwerveDrive().drive(new Translation2d(0, 0), rotateCW / speed / 2, false, false);
    }

  }

  @Override
  public boolean isFinished() {
    return (Math.abs(LimelightHelpers.getTX("limelight-back")) < 1.0) && LimelightHelpers.getTV("limelight-back");
  }

  @Override
  public void end(boolean interrupted) {
    LimelightHelpers.setPipelineIndex("limelight-back", 0);
    swerveSubsystem.getSwerveDrive().drive(new Translation2d(0, 0), 0, false, false);
  }
}

File: src/main/java/frc/robot/commands/drivebase/AimAtLimelightCommand.kt
package frc.robot.commands.drivebase

import edu.wpi.first.math.geometry.Translation2d
import edu.wpi.first.wpilibj.DriverStation
import edu.wpi.first.wpilibj2.command.Command
import frc.robot.subsystems.SwerveSubsystem
import frc.robot.subsystems.VisionSubsystem
import frc.robot.LimelightHelpers.getTV
import frc.robot.LimelightHelpers.getTX
import frc.robot.LimelightHelpers.setPipelineIndex
import kotlin.math.abs

/**
 *
 */
class AimAtLimelightCommand(
    private val swerveSubsystem: SwerveSubsystem,
    private val visionSubsystem: VisionSubsystem
) : Command() {
    private var isRedAlliance = false
    /**
     *
     */
    private var rotateCW: Double = 0.0
    /**
     *
     */
    var speed: Double = 1.0

    init {
        // each subsystem used by the command must be passed into the
        // addRequirements() method (which takes a vararg of Subsystem)
        addRequirements(this.swerveSubsystem, this.visionSubsystem)
    }


    /**
     * The initial subroutine of a command.  Called once when the command is initially scheduled.
     */
    override fun initialize() {
        setPipelineIndex("limelight-back", 1)

        // Take the current angle and see if it's faster to rotate clockwise or counterclockwise to get to 180 degrees
        val currentAngle = swerveSubsystem.heading.degrees // (-180, 180]
        speed = 2.5
        rotateCW = if ((currentAngle > 0)) -speed else speed
    }

    /**
     *
     */
    override fun execute() {
        if (getTV("limelight-back")) {
            swerveSubsystem.swerveDrive.drive(
                Translation2d(0.0, 0.0),
                -visionSubsystem.limelightAimProportionalBack() * 3,
                false,
                false
            )
        } else {
            if (swerveSubsystem.heading.degrees < 140 && swerveSubsystem.heading.degrees > -140) swerveSubsystem.swerveDrive.drive(
                Translation2d(0.0, 0.0),
                rotateCW * 3,
                false,
                false
            )

            else swerveSubsystem.swerveDrive.drive(Translation2d(0.0, 0.0), -rotateCW * 3, false, false)
        }
    }

    /**
     *
     */
    override fun isFinished(): Boolean {
        return (abs(getTX("limelight-back")) < 3.0) && getTV("limelight-back")
    }

    /**
     *
     */
    override fun end(interrupted: Boolean) {
        setPipelineIndex("limelight-back", 0)
        swerveSubsystem.swerveDrive.drive(Translation2d(0.0, 0.0), 0.0, false, false)
    }
}
